WIP: Add preload plugin to compile wasm side modules async

mdboom opened the pull request:
This is a possible solution to #6633 related to dlopen'ing SIDE_MODULES using async wasm compilation (required by Chromium).

This adds a new preload handler for `.so` and `.wasm` files to compile them using the async API (`WebAssembly.instantiate`) at data-loading time.  The actual dynamic linking happens synchronously when calling `dlopen` later.  The downside here is that you may end up compiling wasm that never actually gets dlopen'ed, but I think that's probably a domain-specific problem and will be up to the user to resolve, ultimately.

The only change required on the user end is to add `--use-preload-plugins` when file packaging.

Of note is that `loadWebAssemblyModule` most be called one at a time, since it sets up an environment in global variables that is later used when the wasm is compiled asynchronously.  So this uses a chain of promises to do the compilation, rather than just `Promise.all`.

This will need tests, obviously, but I thought I'd get feedback on the general approach first.

kripken commented on src/library_browser.js:
```diff
@@ -225,6 +225,32 @@ var LibraryBrowser = {
       Module['preloadPlugins'].push(audioPlugin);
 
+#if WASM
+      var wasmPlugin = {};
```
please also put this behind the `MAIN_MODULE` flag, so it's not present in code that doesn't support shared modules.

kripken commented on src/library_browser.js:
```diff
@@ -225,6 +225,32 @@ var LibraryBrowser = {
 
+#if WASM
+      var wasmPlugin = {};
+      wasmPlugin['asyncWasmLoadPromise'] = new Promise((resolve, reject) => resolve());
```
I think `=>` is ES6 notation that currently breaks our JS optimizer. Until we fix that, have to use ES5, sorry...

kripken commented on src/library_browser.js:
```diff
@@ -225,6 +225,32 @@ var LibraryBrowser = {
+      var wasmPlugin = {};
+      wasmPlugin['asyncWasmLoadPromise'] = new Promise((resolve, reject) => resolve());
+      wasmPlugin['canHandle'] = function(name) {
+        return !Module.noWasmDecoding && (name.endsWith('.so') || name.endsWith('.wasm'));
```
what is `noWasmDecoding`?

mdboom replied:
I copied the pattern from the audio and image decoders above.  It's a way for users to turn off the decoder if desired, I suppose.  It might make sense to opt-in rather than opt-out of this, but that would be a change from the other two preload plugins.  Happy to change if it makes sense to.

kripken commented on src/preamble.js:
```diff
@@ -1822,6 +1822,7 @@ function removeRunDependency(id) {
 
 Module["preloadedImages"] = {}; // maps url to image data
 Module["preloadedAudios"] = {}; // maps url to audio data
+Module["preloadedWasm"] = {}; // maps url to wasm instance exports
```
please put this behind the WASM and MAIN_MODULE flags

(also, we should probably put the others behind a flag for preload-plugins, not sure if we have one yet)

kripken commented on src/support.js:
```diff
@@ -86,7 +86,7 @@ function loadDynamicLibrary(lib) {
 #if WASM
 // Loads a side module from binary data
-function loadWebAssemblyModule(binary) {
+function loadWebAssemblyModule(binary, load_async) {
```
`load_async` => `loadAsync`

kripken commented on src/library_browser.js:
```diff
@@ -225,6 +225,33 @@ var LibraryBrowser = {
       };
       Module['preloadPlugins'].push(audioPlugin);
 
+#if WASM && MAIN_MODULE
```
our preprocessor is pretty simple, and doesn't support `&&`... can nest one if inside another.

kripken replied:
(I think that explains the CI errors)

kripken commented on src/library_browser.js:
```diff
@@ -225,6 +225,33 @@ var LibraryBrowser = {
       };
       Module['preloadPlugins'].push(audioPlugin);
 
+#if (WASM != 0) && (MAIN_MODULE != 0)
```
I don't believe this works. Do you see it prevent the code from being included when it should not be?

All I am sure works is
```
#if WASM
#if MAIN_MODULE
...
#endif
#endif
```
Sorry about the ugliness, we should improve the preprocessor (which is literally just a few lines of JS at this point)

kripken commented on src/library_browser.js:
```diff
@@ -225,6 +225,35 @@ var LibraryBrowser = {
+        // promises to run in series.
+        this.asyncWasmLoadPromise = this.asyncWasmLoadPromise.then(
+          function() {
+            return Module.loadWebAssemblyModule(byteArray, true)
```
`loadWebAssemblyModule` should be quoted.

However, does it need to be on `Module` at all? I think you can just call `loadWebAssemblyModule()` here.

kripken replied:
also `;` at end of line

kripken commented on src/library_browser.js:
```diff
@@ -225,6 +225,35 @@ var LibraryBrowser = {
+        // loadWebAssemblyModule can not load modules out-of-order, so rather
+        // than just running the promises in parallel, this makes a chain of
+        // promises to run in series.
+        this.asyncWasmLoadPromise = this.asyncWasmLoadPromise.then(
```
For closure, if a property is quoted once, it must be quoted in all uses, so the two uses on this line must be quoted

mdboom commented:
Quick question about writing a test.  I added a test to `test_browser.py` to load a side module from the virtual file system in this way, but it's failing with the message:

```
using preloaded data can only be done on a web page or in a web worker
```

This is coming from the code that `file_packager.py` generates to test whether it's in a browser.

I had assumed `test_browser.py` tests are only run in the browser.  Is there a way to flag this test to only run in that context?

mdboom commented:
Thanks for the tip.  Making progress.  I have a test mostly working (pushed to this PR).  It displays the expected output in the browser window, but then hangs for a long time and it seems the test harness doesn't get the output.

```
======================================================================
ERROR: test_preload_module (test_browser.browser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdboom/Work/builds/compiling/emscripten/tests/test_browser.py", line 2320, in test_preload_module
    self.btest(main, args=['-s', 'MAIN_MODULE=1', '--preload-file', '.@/', '-O2', '-s', 'WASM=1', '--use-preload-plugins'], expected=expected)
  File "/home/mdboom/Work/builds/compiling/emscripten/tests/runner.py", line 1033, in btest
    self.run_browser(outfile + url_suffix, message, ['/report_result?' + e for e in expected], timeout=timeout)
  File "/home/mdboom/Work/builds/compiling/emscripten/tests/runner.py", line 867, in run_browser
    self.assertIdentical(expectedResult, output)
  File "/home/mdboom/Work/builds/compiling/emscripten/tests/runner.py", line 546, in assertIdentical
    limit_size(''.join([a.rstrip()+'\n' for a in difflib.unified_diff(x.split('\n'), y.split('\n'), fromfile='expected', tofile='actual')]))
Exception: Expected to have '/report_result?hello from main
hello from library' == '[no http server activity]', diff:

--- expected
+++ actual
@@ -1,2 +1 @@
-/report_result?hello from main
-hello from library
+[no http server activity]
```

kripken commented on tests/test_browser.py:
```diff
@@ -2281,6 +2281,44 @@ def test_module(self):
+      #endif
+      }
+    ''')
+    check_execute([PYTHON, EMCC, 'library.c', '-s', 'SIDE_MODULE=1', '-O2', '-o', 'library.wasm', '-s', 'WASM=1'])
```
we prefer `run_process` now for calling python processes

kripken commented:
Ah, another difference is we don't forward stdout/stderr in browser tests. Instead, use the `REPORT_RESULT` macro to report a result code, see e.g. `tests/browser_test_hello_world.c`. That code is what `btest` wants in the `expected` param.

mdboom commented:
Cool.  Thanks for the pointers.  I think I have something that works now...  we'll see how the CI likes it.

kripken commented:
Thanks! Looks good on CI (some other random error there, but doesn't look related).

However, reading the test now, it makes me think maybe a preload plugin on `.wasm` isn't right. Perhaps this should be on `.so` files? Or perhaps `.wasm.so`? Just that `.wasm` files may not all be shared libraries. OTOH there doesn't seem to be any particular suffix proposed in https://github.com/WebAssembly/tool-conventions/blob/master/DynamicLinking.md

kripken commented:
Another thing we should do here is update the docs to mention this. I think we have something about preload-plugins somewhere, but I don't remember offhand.

mdboom commented:
wrt file extensions: `.so` alone would be enough to solve my specific problem.  Let me know where you come down and I'm happy to update the PR.

wrt documentation: [This](https://kripken.github.io/emscripten-site/docs/compiling/Building-Projects.html?highlight=preload#emscripten-ports) is the closest thing I can find in the docs, and even it doesn't go into much detail.  Should I write a preloading file section from scratch?

kripken commented:
Ok, let's do `.so` then.

For docs, yeah - it seems like we don't have a good section on this then. I think the best place to add a subsection is in https://kripken.github.io/emscripten-site/docs/porting/files/packaging_files.html , and please link to it from the place you found, as well as the preloading functions mentioned in https://kripken.github.io/emscripten-site/docs/api_reference/emscripten.h.html#c.emscripten_run_preload_plugins_data Thanks!

mdboom commented:
Great.  I've made this `.so` only and updated the docs.

kripken commented:
Looks good, thanks! Just waiting on CI to re-run (was some problem on the browser run, which is important here).

mdboom commented:
Ah, the test broke because it was testing a SIDE_MODULE with a `.wasm` extension, and this was changed to only handle `.so`.  Hopefully fixed now and it's working for me locally.

kripken commented:
Looks good, thanks!
